File: Medium/question_6.cpp
#include <iostream>

// Abstraction: Abstraction is the process of simplifying complex reality by modeling classes
// based on the essential properties and behaviors. It focuses on what an object does, rather
// than how it does it. Abstraction hides the underlying complexity.
//
// Encapsulation: Encapsulation is the bundling of data (attributes) and methods (functions) that
// operate on the data into a single unit, i.e., a class. It restricts direct access to some of an
// object's components, providing control and security.

class AbstractionExample {
public:
    // Abstraction: The "what" - defining the essential behavior
    void performTask() {
        // Code to perform a task (implementation details are hidden)
    }
};

class EncapsulationExample {
private:
    int data; // Encapsulation: Data is hidden from direct access

public:
    // Encapsulation: The "how" - methods control access to data
    void setData(int value) {
        data = value; // Safely setting the data
    }

    int getData() {
        return data; // Safely retrieving the data
    }
};

File: Medium/question_6.kt

//Abstraction: 
//Abstraction is the process of simplifying complex systems 
//by focusing on essential properties and behaviors while ignoring unnecessary details, 
//allowing for the creation of generalized models or interfaces in programming.

//Encapsulation:
//Encapsulation is the bundling of data and methods that operate on that data into a single unit, 
//typically a class in object-oriented programming, and restricting direct access to the internal 
//details of the object.


// Abstract class representing a Shape

abstract class Shape {
    // Encapsulated property
    private var color: String = ""

    // Abstract method to calculate area
    abstract fun calculateArea(): Double

    // Method to set the color of the shape
    fun setColor(color: String) {
        this.color = color
    }

    // Method to get the color of the shape
    fun getColor(): String {
        return color
    }
}

// Concrete class representing a Circle
class Circle(private val radius: Double) : Shape() {
    override fun calculateArea(): Double {
        return Math.PI * radius * radius
    }
}

fun main() {
    val circle = Circle(5.0)
    circle.setColor("Red")
    println("Circle color: ${circle.getColor()}")
    println("Circle area: ${circle.calculateArea()}")
}
